images/bootstrap/prunedocker/main.go

```diff
@@ -0,0 +1,156 @@
+*/
+
+/*
+   Implements cleaning up unwanted docker data root.
+   This is similar to `docker system prune` but tuned to our needs.
+   TODO(bentheelder): rename this. prune-docker <-> docker prune is a bit lame.
+*/
+
+package main
+
+import (
+	"fmt"
+	"log"
+	"os/exec"
+	"sort"
+	"strings"
+	"time"
+)
+
+var (
+	// images newer than this will be deleted
+	deleteYoungerThanDuration = time.Hour * 24 * 7
+)
+
+// TODO(bentheelder): consider using official docker client library
+// Image represents docker image metadata
+type Image struct {
+	ID         string
+	Repository string
+	Tag        string
+	CreatedAt  time.Time
+}
+
+const (
+	// Docker appears to just use .String() on the timestamps for these
+	// https://golang.org/pkg/time/#Time.String
+	createdAtTimestampFormat = "2006-01-02 15:04:05.999999999 -0700 MST"
+)
+
+func getImages(all bool) ([]Image, error) {
+	args := []string{
+		"images", "--format",
+		// this should contain all fields in Image, comma seperated
+		"{{.ID}},{{.Repository}},{{.Tag}},{{.CreatedAt}}",
+	}
+	if all {
+		args = append(args, "--all")
+	}
+	b, err := exec.Command("docker", args...).Output()
+	if err != nil {
+		return nil, err
+	}
+	images := []Image{}
+	for _, line := range strings.Split(string(b), "\n") {
+		if line == "" {
+			continue
+		}
+		parts := strings.Split(line, ",")
+		if len(parts) != 4 {
+			return nil, fmt.Errorf("failed to parse line: '%v'", line)
+		}
+		parsedTime, err := time.Parse(createdAtTimestampFormat, parts[3])
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse timestamp '%s' %v", parts[1], err)
+		}
+		images = append(images, Image{
+			parts[0],
+			parts[1],
+			parts[2],
+			parsedTime,
+		})
+	}
+	return images, nil
+}
+
+func deleteImage(image Image) error {
+	return exec.Command("docker", "rmi", "-f", image.ID).Run()
+}
+
+func runInShell(command string) error {
+	output, err := exec.Command("/bin/bash", "-c", command).CombinedOutput()
+	log.Printf("%s", output)
+	return err
+}
+
+func removeAllContainers() {
+	runInShell("docker ps -aq | xargs -r docker stop")
+	runInShell("docker ps -aq | xargs -r docker rm")
+}
+
+func removeDanglingImages() {
+	runInShell("docker images --filter dangling=true -qa | xargs -r docker rmi -f")
+}
+
+func pruneVolumes() {
+	runInShell("docker volume prune -f")
+}
+
+func main() {
+	log.Println("Cleaning up Docker data root...")
+	now := time.Now()
+	log.Println("Making sure all containers are stopped / removed.")
+	removeAllContainers()
+	log.Println("Getting images.")
+	// TODO(bentheelder): make sure any lingering containers are removed from the data root
+	// get image refs with metadata
+	images, err := getImages(false)
+	if err != nil {
+		log.Fatalf("Failed to get image data, error: %v", err)
+	}
+	// find images to delete
+	deleteImages := []Image{}
+	for _, image := range images {
+		age := now.Sub(image.CreatedAt)
+		// delete recently created images
+		if age <= deleteYoungerThanDuration ||
```

Comment:
Should we be deleting **all** recently created images? Is there a case where we would want to keep some recently created images around?

Reply:
We only really want to keep images that were pulled (`FROM foo`). Any other images that should be kept should be pushed as an artifact somewhere, not kept on the node.

We can't actually detect if an image was built or pulled so far (you can set the same metadata locally) so we assume any recently created images were built not pulled. This is wrong after switching to a recently released base image, but it will eventually be old enough that this holds.

This should be a reasonable compromise, the problem will be tuning the age we consider recent. If we go too small we'll miss cleaning up sometimes (pod dies after creating an image, no dind jobs manage to clean up on that node for too long), if we go too large on the other hand we're just making more jobs do actual pulls for. One week seemed like a reasonable start. Most base images are months old debian images etc.